File: utils/export.cpp
/**
 * @file export.c
 * @brief Utility to extract individual tensor weights from llama2.c binary model files.
 *
 * Target Data Source:
 *   $ wget https://huggingface.co/karpathy/tinyllamas/resolve/main/stories15M.bin
 */

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>

#define MAKE_DIR(path) mkdir(path, 0755)

typedef struct
{
    int dim;
    int hidden_dim;
    int n_layers;
    int n_heads;
    int n_kv_heads;
    int vocab_size;
    int seq_len;
} Config;

void create_directory_if_not_exists(const char* dir_path)
{
    struct stat st = {0};
    if (stat(dir_path, &st) == -1) {
        if (MAKE_DIR(dir_path) == 0) {
            printf("📂 Creating output directory: %s\n", dir_path);
        }
        else {
            fprintf(stderr, "❌ Failed to create directory: %s\n", dir_path);
            exit(1);
        }
    }
    else {
        printf("📂 Using existing directory: %s\n", dir_path);
    }
}

void save_tensor(const char* output_dir, const char* name, const float* data, int* shape, int ndim)
{
    char filepath[512];
    snprintf(filepath, sizeof(filepath), "%s/%s.bin", output_dir, name);

    FILE* f = fopen(filepath, "wb");
    if (!f) {
        fprintf(stderr, "❌ Error: Cannot write to %s\n", filepath);
        return;
    }

    fwrite(&ndim, sizeof(int), 1, f);

    fwrite(shape, sizeof(int), ndim, f);

    size_t num_elements = 1;
    for (int i = 0; i < ndim; i++) {
        num_elements *= shape[i];
    }
    fwrite(data, sizeof(float), num_elements, f);
    fclose(f);

    printf("💾 Saved: %-25s Shape: [", name);
    for (int i = 0; i < ndim; i++) {
        printf("%d%s", shape[i], (i < ndim - 1) ? ", " : "");
    }
    printf("]\n");
}

void print_usage(const char* program_name)
{
    printf("Usage: %s <model_file> [output_dir]\n\n", program_name);
    printf("Example:\n");
    printf("  %s stories15M.bin my_weights\n", program_name);
    printf("  %s stories15M.bin  (defaults to 'weights')\n\n", program_name);
}

int main(int argc, char* argv[])
{
    if (argc < 2) {
        print_usage(argv[0]);
        return 1;
    }

    if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0) {
        print_usage(argv[0]);
        return 0;
    }

    const char* model_path = argv[1];
    const char* output_dir = (argc >= 3) ? argv[2] : "weights";

    printf("📖 Reading Model: %s\n", model_path);

    FILE* file = fopen(model_path, "rb");
    if (!file) {
        fprintf(stderr, "❌ Cannot open model file: %s\n", model_path);
        return 1;
    }

    Config config;
    if (fread(&config, sizeof(Config), 1, file) != 1) {
        fprintf(stderr, "❌ Failed to read Config.\n");
        fclose(file);
        return 1;
    }

    int shared_weights = config.vocab_size > 0 ? 1 : 0;
    config.vocab_size  = abs(config.vocab_size);

    printf("⚙️  Model Configuration:\n");
    printf("   • dim:        %d\n", config.dim);
    printf("   • hidden_dim: %d\n", config.hidden_dim);
    printf("   • n_layers:   %d\n", config.n_layers);
    printf("   • n_heads:    %d\n", config.n_heads);
    printf("   • n_kv_heads: %d\n", config.n_kv_heads);
    printf(
        "   • vocab_size: %d %s\n", config.vocab_size, shared_weights ? "(Shared)" : "(Unshared)");
    printf("   • seq_len:    %d\n", config.seq_len);
    printf("----------------------------------------\n");

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, sizeof(Config), SEEK_SET);

    long weights_bytes = file_size - sizeof(Config);
    if (weights_bytes <= 0) {
        fprintf(stderr, "❌ File contains no weight data.\n");
        fclose(file);
        return 1;
    }

    float* weights_data = (float*)malloc(weights_bytes);
    if (!weights_data) {
        fprintf(stderr, "❌ Memory allocation failed (%ld bytes).\n", weights_bytes);
        fclose(file);
        return 1;
    }

    printf("🚀 Loading weights into memory...\n");
    if (fread(weights_data, 1, weights_bytes, file) != weights_bytes) {
        fprintf(stderr, "❌ Failed to read weight data.\n");
        free(weights_data);
        fclose(file);
        return 1;
    }
    fclose(file);

    create_directory_if_not_exists(output_dir);
    printf("🚀 Starting extraction...\n");

    float* ptr = weights_data;

    int head_size = config.dim / config.n_heads;

#define SAVE_AND_ADVANCE(name, ndim, ...)            \
    do {                                             \
        int s[] = {__VA_ARGS__};                     \
        save_tensor(output_dir, name, ptr, s, ndim); \
        size_t size = 1;                             \
        for (int i = 0; i < ndim; i++) size *= s[i]; \
        ptr += size;                                 \
    } while (0)

    SAVE_AND_ADVANCE("token_embeddings", 2, config.vocab_size, config.dim);
    SAVE_AND_ADVANCE("layers_rms_att_weight", 2, config.n_layers, config.dim);
    SAVE_AND_ADVANCE("layers_wq", 3, config.n_layers, config.dim, config.n_heads * head_size);
    SAVE_AND_ADVANCE("layers_wk", 3, config.n_layers, config.dim, config.n_kv_heads * head_size);
    SAVE_AND_ADVANCE("layers_wv", 3, config.n_layers, config.dim, config.n_kv_heads * head_size);
    SAVE_AND_ADVANCE("layers_wo", 3, config.n_layers, config.n_heads * head_size, config.dim);
    SAVE_AND_ADVANCE("layers_rms_ffn_weight", 2, config.n_layers, config.dim);
    SAVE_AND_ADVANCE("layers_w1", 3, config.n_layers, config.dim, config.hidden_dim);
    SAVE_AND_ADVANCE("layers_w2", 3, config.n_layers, config.hidden_dim, config.dim);
    SAVE_AND_ADVANCE("layers_w3", 3, config.n_layers, config.dim, config.hidden_dim);
    SAVE_AND_ADVANCE("final_rms_norm", 1, config.dim);

    if (shared_weights) {
        printf("ℹ️  Shared Weights detected: Copying token_embeddings to output_wcls...\n");
        int s[] = {config.vocab_size, config.dim};
        save_tensor(output_dir, "output_wcls", weights_data, s, 2);
    }
    else {
        SAVE_AND_ADVANCE("output_wcls", 2, config.vocab_size, config.dim);
    }

    free(weights_data);
    printf("\n✨ Done! All weights have been extracted to '%s'.\n", output_dir);

    return 0;
}

File: utils/export.py
# /**
#  * @file export.py
#  * @brief Utility to extract individual tensor weights from llama2.c binary model files.
#  *
#  * Target Data Source:
#  *   $ wget https://huggingface.co/karpathy/tinyllamas/resolve/main/stories110M.bin
#  *    python export.py stories110M.bin stories110M
#  */

import sys
import struct
import os
import numpy as np

def create_directory_if_not_exists(dir_path):
    if not os.path.exists(dir_path):
        try:
            os.makedirs(dir_path, mode=0o755)
            print(f"📂 Creating output directory: {dir_path}")
        except OSError as _:
            print(f"❌ Failed to create directory: {dir_path}")
            sys.exit(1)
    else:
        print(f"📂 Using existing directory: {dir_path}")

def save_tensor(output_dir, name, data):
    """
    保存 Tensor 为二进制文件，格式与 C 代码一致：
    [ndim (int), shape (int...), data (float...)]
    """
    filepath = os.path.join(output_dir, f"{name}.bin")
    
    shape = data.shape
    ndim = len(shape)
    
    with open(filepath, 'wb') as f:
        # 1. 写入 ndim
        f.write(struct.pack('i', ndim))
        # 2. 写入 shape
        f.write(struct.pack(f'{ndim}i', *shape))
        # 3. 写入数据 (确保是 float32)
        data.astype(np.float32).tofile(f)

    # 打印日志
    shape_str = ", ".join(map(str, shape))
    print(f"💾 Saved: {name:<25} Shape: [{shape_str}]")

def main():
    if len(sys.argv) < 2:
        print(f"Usage: python {sys.argv[0]} <model_file> [output_dir]")
        sys.exit(1)

    model_path = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) >= 3 else "weights"

    print(f"📖 Reading Model: {model_path}")

    if not os.path.isfile(model_path):
        print(f"❌ Cannot open model file: {model_path}")
        sys.exit(1)

    # 读取 Config (7个 int)
    # C struct: dim, hidden_dim, n_layers, n_heads, n_kv_heads, vocab_size, seq_len
    with open(model_path, 'rb') as f:
        config_bytes = f.read(28) # 7 * 4 bytes
        if len(config_bytes) != 28:
            print("❌ Failed to read Config.")
            sys.exit(1)
        
        config = struct.unpack('iiiiiii', config_bytes)
        dim, hidden_dim, n_layers, n_heads, n_kv_heads, vocab_size, seq_len = config

        # 处理 Shared Weights 标志
        shared_weights = vocab_size > 0
        vocab_size = abs(vocab_size)

        print("⚙️  Model Configuration:")
        print(f"   • dim:        {dim}")
        print(f"   • hidden_dim: {hidden_dim}")
        print(f"   • n_layers:   {n_layers}")
        print(f"   • n_heads:    {n_heads}")
        print(f"   • n_kv_heads: {n_kv_heads}")
        print(f"   • vocab_size: {vocab_size} {'(Shared)' if shared_weights else '(Unshared)'}")
        print(f"   • seq_len:    {seq_len}")
        print("----------------------------------------")

        # 读取剩余的所有权重数据
        print("🚀 Loading weights into memory...")
        # 从当前位置读取剩余所有数据作为 float32 数组
        weights_data = np.fromfile(f, dtype=np.float32)

    if weights_data.size == 0:
        print("❌ File contains no weight data.")
        sys.exit(1)

    create_directory_if_not_exists(output_dir)
    print("🚀 Starting extraction...")

    # 指针偏移量
    offset = 0
    head_size = dim // n_heads

    # 定义需要转置的权重名称集合
    transpose_targets = {
        "layers_wq", "layers_wk", "layers_wv", "layers_wo",
        "layers_w1", "layers_w2", "layers_w3", "output_wcls"
    }

    # 辅助函数：提取并处理权重
    token_embeddings_ref = None # 用于 shared weights

    def extract_and_save(name, shape):
        nonlocal offset, token_embeddings_ref
        
        # 计算元素总数
        size = np.prod(shape)
        
        # 切片
        data = weights_data[offset : offset + size]
        offset += size
        
        # Reshape
        data = data.reshape(shape)

        # 保存 token_embeddings 的引用，以防 output_wcls 需要共享
        if name == "token_embeddings":
            token_embeddings_ref = data

        # 检查是否需要转置 (最后两个维度)
        if name in transpose_targets:
            # 交换最后两个维度 (-1 和 -2)
            data = np.swapaxes(data, -1, -2)
            # 注意：转置后为了保证内存连续性以便正确写入文件，通常需要 contiguous()
            # 但 numpy 的 tofile 会自动处理，或者我们可以显式调用
            data = np.ascontiguousarray(data)

        save_tensor(output_dir, name, data)

    # --- 按顺序提取权重 ---

    # 1. token_embeddings
    extract_and_save("token_embeddings", (vocab_size, dim))

    # 2. layers_rms_att_weight
    extract_and_save("layers_rms_att_weight", (n_layers, dim))

    # 3. layers_wq
    extract_and_save("layers_wq", (n_layers, dim, n_heads * head_size))

    # 4. layers_wk
    extract_and_save("layers_wk", (n_layers, dim, n_kv_heads * head_size))

    # 5. layers_wv
    extract_and_save("layers_wv", (n_layers, dim, n_kv_heads * head_size))

    # 6. layers_wo
    extract_and_save("layers_wo", (n_layers, n_heads * head_size, dim))

    # 7. layers_rms_ffn_weight
    extract_and_save("layers_rms_ffn_weight", (n_layers, dim))

    # 8. layers_w1
    extract_and_save("layers_w1", (n_layers, hidden_dim, dim))

    # 9. layers_w2
    extract_and_save("layers_w2", (n_layers, dim, hidden_dim))

    # 10. layers_w3
    extract_and_save("layers_w3", (n_layers, hidden_dim, dim))

    # 11. final_rms_norm
    extract_and_save("final_rms_norm", (dim,))

    # 12. output_wcls
    if shared_weights:
        print("ℹ️  Shared Weights detected: Copying token_embeddings to output_wcls...")
        # 复制 token_embeddings
        data = token_embeddings_ref.copy()
        
        # 即使是共享的，按照你的要求，output_wcls 也需要转置
        # token_embeddings 是 [vocab, dim]，转置后变成 [dim, vocab]
        if "output_wcls" in transpose_targets:
            data = np.swapaxes(data, -1, -2)
            data = np.ascontiguousarray(data)
            
        save_tensor(output_dir, "output_wcls", data)
    else:
        extract_and_save("output_wcls", (vocab_size, dim))

    print(f"\n✨ Done! All weights have been extracted to '{output_dir}'.")

if __name__ == "__main__":
    main()